### DISCUSSION TITLE
unocss In order to load dynamic css variables, index.html adds @unocss/runtime

### DISCUSSION THREAD

**Discussion by Flame-00** [unocss/unocss#3310] [Q&A]
### UnoCSS version

0.56.1

### Describe the bug

This is the cdn file I introduced in index.html
![image](https://github.com/unocss/unocss/assets/101857530/bc6f83e8-4614-41ca-b262-e41074204391)
Download @unocss/runtime from the diagram
![image](https://github.com/unocss/unocss/assets/101857530/a7302fc4-3ce2-404f-b549-f4cc7e751f70)
这是我在main引入的npm下载的 未生效
![image](https://github.com/unocss/unocss/assets/101857530/3065f6f8-1a70-4196-9896-8582f2d54574)

There's another problem
In addition, there is another problem, because the document says that unocss is mounted after dom loading is completed, and I have been making a requirement recently, which is to obtain the width and height of the element dynamically after dom mounting, but I find that the padding provided by unocss does not take effect in time, resulting in the real width and height obtained does not include the width and height of unocss. This is not the case when the page is refreshed, only when the project is launched for the first time to enter the component for the first time, how to solve this asynchronous problem?
![image](https://github.com/unocss/unocss/assets/101857530/944375cd-11da-4c36-8557-467948df82df)
![image](https://github.com/unocss/unocss/assets/101857530/7332eb9b-6792-49b7-8efb-d848feceec2c)

Note that refreshing the page padding is available, but the main error is restarting the project on the current page and then remounting unocss will cause the asynchronous problem described above!

![image](https://github.com/unocss/unocss/assets/101857530/01f25e26-a1ca-429b-b158-b2764d4545e1)

### Reproduction

This is the cdn file I introduced in index.html
![image](https://github.com/unocss/unocss/assets/101857530/bc6f83e8-4614-41ca-b262-e41074204391)
Download @unocss/runtime from the diagram
![image](https://github.com/unocss/unocss/assets/101857530/a7302fc4-3ce2-404f-b549-f4cc7e751f70)
这是我在main引入的npm下载的 未生效
![image](https://github.com/unocss/unocss/assets/101857530/3065f6f8-1a70-4196-9896-8582f2d54574)

There's another problem
In addition, there is another problem, because the document says that unocss is mounted after dom loading is completed, and I have been making a requirement recently, which is to obtain the width and height of the element dynamically after dom mounting, but I find that the padding provided by unocss does not take effect in time, resulting in the real width and height obtained does not include the width and height of unocss. This is not the case when the page is refreshed, only when the project is launched for the first time to enter the component for the first time, how to solve this asynchronous problem?
![image](https://github.com/unocss/unocss/assets/101857530/944375cd-11da-4c36-8557-467948df82df)
![image](https://github.com/unocss/unocss/assets/101857530/7332eb9b-6792-49b7-8efb-d848feceec2c)

Note that refreshing the page padding is available, but the main error is restarting the project on the current page and then remounting unocss will cause the asynchronous problem described above!

![image](https://github.com/unocss/unocss/assets/101857530/01f25e26-a1ca-429b-b158-b2764d4545e1)

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/unocss/unocss/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't [already an issue](https://github.com/unocss/unocss/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/unocss/unocss/discussions) or join our [Discord Chat Server](https://chat.antfu.me).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by zyyv** [comment]
What's reason cause you use runtime package. I use unocss vite plugin mostly.

**Comment by Flame-00** [comment]
> 是什么原因导致您使用运行时包。我主要使用unocss vite插件。

Because if you use the template string variable

**Comment by zyyv** [comment]
你可以分享一个最小复现的例子（仓库或 stackblitz 链接）吗，因为动态的 css 我们一般建议提前放入 `safelist` 配置项中，不太清楚你的场景。我觉得runtime package 使用的场景比较少 :( 个人觉得。

**Reply by Flame-00** [reply]
![image](https://github.com/unocss/unocss/assets/101857530/67a0fb5b-90f5-4d89-b303-ee4fafda2633)
It's a very basic variable or something a little more complicated

**Reply by Flame-00** [reply]
Because what I'm thinking is that if you put it all in the safelist, then every time you write a dynamic variable you have to add a new rule to the safelist, which is a little cumbersome

**Reply by Flame-00** [reply]
Hello, I do not understand the safelist play, can you give a few more examples, or shallow teach me?

**Reply by zyyv** [reply]
> which is a little cumbersome

Yep, but uno is static template extraction

Abandoning the use of vite plugin and using runtime may have higher overhead and is not a good solution. It may be better to put it in `safelist`.

**Reply by zyyv** [reply]
```ts
// uno.config.ts
{
// ...
  safelist: [
    ...Array.from({ length: 5 }, (_, i) => `grid-cols-${i + 1}`),
  ]
}
```
like this, the principle is to generate the utility you need to use in advance.

Refer to https://unocss.dev/guide/extracting#safelist.

**Reply by Flame-00** [reply]
Hello, I may be stupid, if the gridCount is a single number from the parent component, rather than an array, how to write it?

**Reply by Flame-00** [reply]
Hello, the second problem mentioned in the article is my headache, because unocss style is always mounted after dom generation, which causes me trouble when I get a dynamic width and height, I have to use the traditional style style. Is there a plan for this?

**Reply by zyyv** [reply]
> Hello, I may be stupid, if the gridCount is a single number from the parent component, rather than an array, how to write it?

I mean, write all of the generated results in `safelist`.

> because unocss style is always mounted after dom generation

emm... I think uno simply injects css and does not affect the mounting of dom. In your screenshot, using nextTick is enough to get the latest dom attributes.

**Reply by Flame-00** [reply]
You may misunderstand what I mean, now the situation is that unocss can get the new dom properties after nextTick, but if the service is stopped in this page, pnpm dev unocss will not work properly. Asynchronous problems can occur unless a timer of several seconds is added. Otherwise we can't fix it.
